plugin/undotree.vim

```diff
@@ -148,1306 +148,17 @@ if !exists('g:undotree_HelpLine')
-            for j in w:undotree_diffmatches
-                call matchdelete(j)
-            endfor
-            let w:undotree_diffmatches = []
-        endif
-    endfor
-
-    "restore position
-    call s:exec_silent("norm! ".curwinnr."\<c-w>\<c-w>")
-    call winrestview(savedview)
-endfunction
-
-function! s:diffpanel.Hide()
-    call s:log(self.bufname." Hide()")
-    if !self.IsVisible()
-        return
-    endif
-    call self.SetFocus()
-    call s:exec("quit")
-    call self.CleanUpHighlight()
-endfunction
-
-"=================================================
-" It will set the target of undotree window to the current editing buffer.
-function! s:undotreeAction(action)
-    call s:log("undotreeAction()")
-    if !exists('t:undotree')
-        echoerr "Fatal: t:undotree does not exist!"
-        return
-    endif
-    call t:undotree.Action(a:action)
-endfunction
-
-function! s:exitIfLast()
-    let num = 0
-    if t:undotree.IsVisible()
-        let num = num + 1
-    endif
-    if t:diffpanel.IsVisible()
-        let num = num + 1
-    endif
-    if winnr('$') == num
-        call t:undotree.Hide()
-        call t:diffpanel.Hide()
-    endif
-endfunction
-
-"=================================================
-"called outside undotree window
-function! UndotreeUpdate()
-    if !exists('t:undotree')
-        return
-    endif
-    if !exists('w:undotree_id')
-        let w:undotree_id = 'id_'.s:getUniqueID()
-        call s:log("Unique window id assigned: ".w:undotree_id)
-    endif
-    " assume window layout won't change during updating.
-    let thiswinnr = winnr()
-    call t:undotree.Update()
-    " focus moved
-    if winnr() != thiswinnr
-        call s:exec("norm! ".thiswinnr."\<c-w>\<c-w>")
-    endif
-endfunction
-
-function! UndotreeToggle()
-    call s:log(">>> UndotreeToggle()")
-    if !exists('w:undotree_id')
-        let w:undotree_id = 'id_'.s:getUniqueID()
-        call s:log("Unique window id assigned: ".w:undotree_id)
-    endif
-
-    if !exists('t:undotree')
-        let t:undotree = s:new(s:undotree)
-        let t:diffpanel = s:new(s:diffpanel)
-    endif
-    call t:undotree.Toggle()
-    call s:log("<<< UndotreeToggle() leave")
-endfunction
-
-function! UndotreeIsVisible()
-    return (exists('t:undotree') && t:undotree.IsVisible())
-endfunction
-
-function! UndotreeHide()
-    if UndotreeIsVisible()
-        call UndotreeToggle()
-    endif
-endfunction
-
-function! UndotreeShow()
-    if ! UndotreeIsVisible()
-        call UndotreeToggle()
-    else
-        call t:undotree.SetFocus()
-    endif
-endfunction
-
-function! UndotreeFocus()
-    if UndotreeIsVisible()
-        call t:undotree.SetFocus()
-    endif
-endfunction
-
-
 "let s:auEvents = "InsertEnter,InsertLeave,WinEnter,WinLeave,CursorMoved"
 let s:auEvents = "BufEnter,InsertLeave,CursorMoved,BufWritePost"
 augroup Undotree
-    exec "au! ".s:auEvents." * call UndotreeUpdate()"
+    exec "au! ".s:auEvents." * call undotree#UndotreeUpdate()"
```

Comment:
CursorMoved and BufEnter are always triggered, so this defeats the purpose of `autoload/`, which is supposed to only load a plugin if the user actually needs it.

Reply:
Yes, but [these lines](https://github.com/benknoble/undotree/blob/a707a649fac743c34ad17ef16509a4076a35116a/autoload/undotree.vim#L1257) show that that function is quick to return if necessary. Granted, it still forces a loading of the file, and there might be a better way to do so.

Reply:
indeed. currently I'm not sure if I can get rid of the event because that's how undotree handle inputs and update multiple buffers.
I'll leave it there until we figure out a better solution.